lighthouse-core/report/html/renderer/element-screenshot-renderer.js

```diff
@@ -0,0 +1,237 @@
+            <polygon points="0,${bottom} 1,${bottom} 1,1 0,1" ></polygon>
+            <polygon points="0,${top} ${left},${top} ${left},${bottom} 0,${bottom}" ></polygon>
+            <polygon points="${right},${top} 1,${top} 1,${bottom} ${right},${bottom}" ></polygon>
+          </clipPath>
+        </defs>
+      </svg>`;
+    return clipPathSvg;
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {LH.Artifacts.FullPageScreenshot} fullPageScreenshot
+   */
+  static _installBackgroundImageStyle(dom, fullPageScreenshot) {
+    const containerEl = dom.find('.lh-container', dom._document);
+    if (containerEl.querySelector('#full-page-screenshot-style')) return;
+
+    const fullpageScreenshotUrl = fullPageScreenshot.data;
+    const fullPageScreenshotStyleTag = dom.createElement('style');
+    fullPageScreenshotStyleTag.id = 'full-page-screenshot-style';
+    fullPageScreenshotStyleTag.innerText = `
+      .lh-element-screenshot__image {
+        background-image: url(${fullpageScreenshotUrl})
+      }`;
+    containerEl.appendChild(fullPageScreenshotStyleTag);
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {ParentNode} templateContext
+   * @param {LH.Artifacts.FullPageScreenshot} fullPageScreenshot
+   */
+  static installOverlayFeature(dom, templateContext, fullPageScreenshot) {
+    ElementScreenshotRenderer._installBackgroundImageStyle(dom, fullPageScreenshot);
+
+    for (const el of dom.document().querySelectorAll('.lh-element-screenshot')) {
+      el.addEventListener('click', () => {
+        const overlay = dom.createElement('div');
+        overlay.classList.add('lh-element-screenshot__overlay');
+        const clipRect = {
+          width: Number(el.getAttribute('rectWidth')),
+          height: Number(el.getAttribute('rectHeight')),
+          left: Number(el.getAttribute('rectLeft')),
+          right: Number(el.getAttribute('rectRight')),
+          top: Number(el.getAttribute('rectTop')),
+          bottom: Number(el.getAttribute('rectBottom')),
+        };
+        overlay.appendChild(ElementScreenshotRenderer.render(
+          dom,
+          templateContext,
+          clipRect,
+          fullPageScreenshot,
+          {
+            // TODO: should this be documentElement width?
+            width: window.innerWidth * 0.75,
+            height: window.innerHeight * 0.75,
+          }
+        ));
+        document.body.appendChild(overlay);
+        overlay.addEventListener('click', () => {
+          overlay.remove();
+        });
+      });
+    }
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {ParentNode} templateContext
+   * @param {LH.Artifacts.Rect} clipRect Region of screenshot to highlight.
+   * @param {LH.Artifacts.FullPageScreenshot} fullPageScreenshot
+   * @param {Size} viewportSize Size of region to render screenshot in.
+   * @return {Element}
+   */
+  static render(dom, templateContext, clipRect, fullPageScreenshot, viewportSize) {
```

Comment:
this function could really use some explainin' of the whole viewport thing or breaking up or both 😅 

EDIT: ok now that I've gone through everything I understand where viewport was coming from. I think `containerSize` and `imageViewport` instead of `viewportSize` and `viewport` or something might be my personal favorite, but *something* to clarify the viewport isn't the page viewport would be nice